File: find-first-and-last-position-of-element-in-sorted-array.java
//Time: O(logn) and Space: O(n)

class Solution {
    private static int binarySearchFirst(int nums[], int target) {
        int left = 0;
        int high = nums.length - 1;
        
        while (left <= high) {
            int mid = left + (high - left) / 2;
            
            if (nums[mid] == target) {
                if (mid == 0 || nums[mid] > nums[mid - 1]) {
                    return mid;
                } else {
                    high = mid - 1;
                }
            } else if (nums[mid] < target) {
                left = mid + 1;
            } else {
                high = mid - 1;
            }
        }
        
        return -1;
    }

    private static int binarySearchLast(int nums[], int target) {
        int left = 0;
        int high = nums.length - 1;
        
        while (left <= high) {
            int mid = left + (high - left) / 2;
            
            if (nums[mid] == target) {
                if (mid == nums.length - 1 || nums[mid] < nums[mid + 1]) {
                    return mid;
                } else {
                    left = mid + 1;
                }
            } else if (nums[mid] > target) {
                high = mid - 1;
            } else {
                left = mid + 1;
            }
        }
        
        return -1;
    }

    public int[] searchRange(int[] nums, int target) {
        if (nums == null || nums.length == 0) {
            return new int[]{-1, -1};
        }
        
        int n = nums.length;
        if (nums[0] > target || nums[n - 1] < target) {
            return new int[]{-1, -1};
        }
        
        int first = binarySearchFirst(nums, target);
        int last = binarySearchLast(nums, target);
        
        return new int[]{first, last};
    }
}

File: FindFirstAndLastPositionOfElementInSortedArray.py
class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        
        
        # If the element is not in the array then return [-1,-1] directly
        if len(nums) == 0:
            return [-1,-1]
        
        if (target > nums[-1]) or (target < nums[0]):
            return [-1,-1]
        
        # function for getting the first index
        def BinarySearchFirstIndex(nums, target):
            left = 0
            right = len(nums) - 1
            
            while left <= right:
                mid = (left + right) // 2
                
                if nums[mid] == target: # if target is found
                    # Then we check whether the mid is first index
                    # if it is then return the mid because it will be the first occurance
                    # or else we check if the immediate left element is less then the current
                    # this will prove that we are at the first occurance of target
                    if ((mid == 0) or (nums[mid - 1] < nums[mid])):
                        return mid
                    right = mid - 1
                elif (nums[mid] < target):
                    left = mid + 1
                elif (nums[mid] > target):
                    right = mid - 1
            return -1
        
        def BinarySearchLastIndex(nums, target):
            left = 0
            right = len(nums) - 1
            
            while left <= right:
                mid = (left + right) // 2
                
                if nums[mid] == target: # if target is found
                    # Then we check whether the mid is at last index
                    # If it is then return the mid because it will be the last occurance
                    # Or else we check if the immediate right element is greater then the current
                    # This will prove that we are at the last occurance of target
                    if ((mid == len(nums) - 1) or (nums[mid + 1] > nums[mid])):
                        return mid
                    left = mid + 1
                elif (nums[mid] < target):
                    left = mid + 1
                elif (nums[mid] > target):
                    right = mid - 1
            return -1
                
        firstIndex = BinarySearchFirstIndex(nums, target)
        secondIndex = BinarySearchLastIndex(nums, target)
        
        return [firstIndex, secondIndex]
            
        # Time Complexity: O(log(n))
        # Space Complexity: O(1)
            